### PR TITLE
W-15226975: Core scopes and routers must be Operations and not Constructs

### PR DESCRIPTION
**Pull request by mbuchwald**

### CODE REVIEW

**Comment by mbuchwald** [comment]
--validate

**Comment by mbuchwald** [comment]
--validate

**Review by elrodro83 — DISMISSED** [review]

File: modules/extensions-mule-support/src/main/java/org/mule/runtime/module/extension/mule/internal/loader/parser/MuleSdkOperationModelParser.java [hunk 1]
```diff
@@ -403,14 +403,14 @@ private Stream<ComponentAstWithHierarchy> getOperationsAstRecursiveStream(Map<St
 
           if (filtered.contains(componentAst) || ignoreCondition.test(componentAstWithHierarchy)) {
             return Stream.empty();
+          } else if (filterCondition.test(componentAstWithHierarchy)) {
```
Comment (marianogonzalez) [inline modules/extensions-mule-support/src/main/java/org/mule/runtime/module/extension/mule/internal/loader/parser/MuleSdkOperationModelParser.java:406]: why this change? isn't this altering behavior? With the prior logic, it'd be possible for a component to be returned even though the filtered condition would also be true

File: runtime-extension-model/src/main/java/org/mule/runtime/core/api/extension/provider/MuleExtensionModelDeclarer.java [hunk 2]
```diff
@@ -798,14 +817,18 @@ private void declareScatterGather(ExtensionDeclarer extensionDeclarer, ClassType
         .withLayout(LayoutModel.builder().tabName(ADVANCED_TAB).build())
         .withModelProperty(new TargetModelProperty());
 
+    // Comment to remember for the PR: how to set array of any?
```
Comment (marianogonzalez) [inline runtime-extension-model/src/main/java/org/mule/runtime/core/api/extension/provider/MuleExtensionModelDeclarer.java:820]: ArrayType with AnyType as the value

**Review by marianogonzalez — CHANGES_REQUESTED** [review]

File: modules/extensions-mule-support/src/main/java/org/mule/runtime/module/extension/mule/internal/loader/parser/MuleSdkOperationModelParser.java (hunk 1 above)
Comment (mbuchwald) [reply modules/extensions-mule-support/src/main/java/org/mule/runtime/module/extension/mule/internal/loader/parser/MuleSdkOperationModelParser.java:406]: The problem is that before a scope/router that should have been ignored won't be ignored. Since in the next line we check if there is an operation model, and then we proceed... but now the scopes and routers are operations, so we find an operation model for them, and we don't ignore them. 
The example is for checking if an operation is transactional, and it has an async which inside has a transactional try. That operation is not transactional, and we have this into consideration in the ignore corresponding to isTransactional.. but since now there is an operation model, we don't check if it has to be ignored

Comment (mbuchwald) [reply modules/extensions-mule-support/src/main/java/org/mule/runtime/module/extension/mule/internal/loader/parser/MuleSdkOperationModelParser.java:406]: And that logic was added in particular for cases like this one :/

File: runtime-extension-model/src/main/java/org/mule/runtime/core/api/extension/provider/MuleExtensionModelDeclarer.java [hunk 3]
```diff
@@ -849,13 +873,16 @@ private void declareParallelForEach(ExtensionDeclarer extensionDeclarer, ClassTy
         .withDisplayModel(DisplayModel.builder().displayName(TARGET_VALUE_PARAMETER_DISPLAY_NAME).build())
         .withLayout(LayoutModel.builder().tabName(ADVANCED_TAB).build())
         .withModelProperty(new TargetModelProperty());
+
+    parallelForeach.withOutput().ofDynamicType(BaseTypeBuilder.create(MetadataFormat.JAVA).arrayType().of(ANY_TYPE).build());
+    parallelForeach.withOutputAttributes().ofDynamicType(ANY_TYPE);
```
Comment (marianogonzalez) [inline runtime-extension-model/src/main/java/org/mule/runtime/core/api/extension/provider/MuleExtensionModelDeclarer.java:878]: is it any or is it void? does it retain whatever attributes entered the parallel-foreach or does it nullify it?

Comment (mbuchwald) [reply runtime-extension-model/src/main/java/org/mule/runtime/core/api/extension/provider/MuleExtensionModelDeclarer.java:878]: As far as I know, it keeps the attributes and combines them in case there is more than one

**Comment by mbuchwald** [comment]
--validate

**Review by marianogonzalez — APPROVED** [review]
